rest/diagnostic_doc_api.go

```diff
@@ -70,20 +76,89 @@ func (h *handler) handleGetDocChannels() error {
+	// Only require a valid JSON payload if docid is not provided.
+	// If docid is provided, the sync function will use the document from the bucket, and the payload is optional.
 	if err != nil {
-		if docid == "" {
-			return fmt.Errorf("no doc id provided for dry run and error reading body: %s", err)
+		return base.HTTPErrorf(http.StatusBadRequest, "Error reading sync function payload: %v", err)
+	}
+
+	if syncDryRunPayload.Doc == nil && docid == "" {
+		return base.HTTPErrorf(http.StatusBadRequest, "no docid or document provided")
+	}
+
+	oldDoc := &db.Document{ID: docid}
+	oldDoc.UpdateBody(syncDryRunPayload.Doc)
+	if docid != "" {
+		if docInbucket, err := h.collection.GetDocument(h.ctx(), docid, db.DocUnmarshalAll); err == nil {
+			oldDoc = docInbucket
+			if len(syncDryRunPayload.Doc) == 0 {
+				syncDryRunPayload.Doc = oldDoc.Body(h.ctx())
+				oldDoc.UpdateBody(nil)
+			}
+		} else {
+			return base.HTTPErrorf(http.StatusNotFound, "Error reading document: %v", err)
 		}
+	} else {
+		oldDoc.UpdateBody(nil)
+	}
+
+	delete(syncDryRunPayload.Doc, db.BodyId)
+
+	// Get the revision ID to match, and the new generation number:
+	matchRev, _ := syncDryRunPayload.Doc[db.BodyRev].(string)
+	generation, _ := db.ParseRevID(h.ctx(), matchRev)
+	if generation < 0 {
+		return base.HTTPErrorf(http.StatusBadRequest, "Invalid revision ID")
+	}
+	generation++
+
+	// Create newDoc which will be used to pass around Body
+	newDoc := &db.Document{
+		ID: docid,
 	}
+	// Pull attachments
+	newDoc.SetAttachments(db.GetBodyAttachments(syncDryRunPayload.Doc))
+	delete(syncDryRunPayload.Doc, db.BodyAttachments)
+	delete(syncDryRunPayload.Doc, db.BodyRevisions)
 
-	output, err, syncFnErr := h.collection.SyncFnDryrun(h.ctx(), body, docid)
+	if _, ok := syncDryRunPayload.Doc[base.SyncPropertyName]; ok {
+		return base.HTTPErrorf(http.StatusBadRequest, "document-top level property '_sync' is a reserved internal property")
+	}
+
+	db.StripInternalProperties(syncDryRunPayload.Doc)
+
+	// We needed to keep _deleted around in the body until we generate rev ID, but now it can be removed
+	_, isDeleted := syncDryRunPayload.Doc[db.BodyDeleted]
+	if isDeleted {
+		delete(syncDryRunPayload.Doc, db.BodyDeleted)
+	}
+
+	//update the newDoc body to be without any special properties
+	newDoc.UpdateBody(syncDryRunPayload.Doc)
+
+	rawDocBytes, err := newDoc.BodyBytes(h.ctx())
 	if err != nil {
-		return err
+		return base.HTTPErrorf(http.StatusBadRequest, "Error marshalling document: %v", err)
 	}
-	if syncFnErr != nil {
+
+	newRev := db.CreateRevIDWithBytes(generation, matchRev, rawDocBytes)
+	newDoc.RevID = newRev
+
+	output, err := h.collection.SyncFnDryrun(h.ctx(), newDoc, oldDoc, syncDryRunPayload.Function)
+	if err != nil {
+		var syncFnDryRunErr *base.SyncFnDryRunError
+		if !errors.As(err, &syncFnDryRunErr) {
+			return err
+		}
+
+		errMsg := syncFnDryRunErr.Error()
+		if syncFnDryRunErr.Unwrap() != nil {
+			errMsg = syncFnDryRunErr.Err.Error()
+		}
```

Comment:
Stole comment from the import PR since I spotted it there first.

---

I don't think the `Unwrap` step isn't necessary since `Error()` on the typed error returns the underlying error (if present) - and if we have a nil typed error, we can use that without a problem since it returns `syncFnDryRunErrorPrefix` as a generic error message.


```suggestion
		errMsg := syncFnDryRunErr.Error()
```